File: desktop/frontend/src/registry/registry.test.ts
import { describe, expect, it } from 'vitest'
import type { SessionPrincipal } from '../auth/session'
import { applyEffectiveModuleEnablement, isPathModuleEnabled, moduleEnablementRegistry, resetEffectiveModuleEnablement } from './moduleEnablement'
import { canAccessNavigationPath, authenticatedNavigationRegistry } from './navigation'
import { moduleRegistry } from './modules'
import { permissionRegistry } from './permissions'

function principalWith(permissions: string[], roles: string[] = ['Admin']): SessionPrincipal {
  return {
    id: 1,
    username: 'tester',
    roles,
    permissions,
  }
}

describe('desktop registries', () => {
  it('defines module enablement defaults for baseline platform and Sukumad modules', () => {
    expect(moduleEnablementRegistry.map((item) => item.moduleId)).toEqual([
      'dashboard',
      'administration',
      'settings',
      'servers',
      'requests',
      'deliveries',
      'jobs',
      'scheduler',
      'observability',
      'documentation',
      'orgunits',
      'reporters',
    ])
    expect(moduleEnablementRegistry.every((item) => item.enabledByDefault)).toBe(true)
  })

  it('defines unique permission keys and includes platform and Sukumad permissions', () => {
    const keys = permissionRegistry.map((item) => item.key)
    expect(new Set(keys).size).toBe(keys.length)
    expect(keys).toEqual(
      expect.arrayContaining([
        'users.read',
        'users.write',
        'audit.read',
        'settings.read',
        'settings.write',
        'servers.read',
        'requests.read',
        'deliveries.read',
        'jobs.read',
        'scheduler.read',
        'observability.read',
        'orgunits.read',
        'reporters.read',
      ]),
    )
  })

  it('defines baseline modules for platform and Sukumad placeholders', () => {
    const moduleIds = moduleRegistry.map((item) => item.id)
    expect(moduleIds).toEqual([
      'dashboard',
      'administration',
      'settings',
      'servers',
      'requests',
      'deliveries',
      'jobs',
      'scheduler',
      'observability',
      'documentation',
      'orgunits',
      'reporters',
    ])
    expect(moduleRegistry.find((item) => item.id === 'administration')?.navItems).toEqual([
      'users',
      'roles',
      'permissions',
      'audit',
      'settings',
    ])
  })

  it('defines grouped navigation with administration and Sukumad children', () => {
    const admin = authenticatedNavigationRegistry.find((item) => item.id === 'administration')
    expect(admin?.children?.map((item) => item.id)).toEqual(['users', 'roles', 'permissions', 'audit', 'settings'])
    const sukumad = authenticatedNavigationRegistry.find((item) => item.id === 'sukumad')
    expect(sukumad?.children?.map((item) => item.id)).toEqual(['servers', 'requests', 'deliveries', 'jobs', 'scheduler', 'observability', 'documentation', 'orgunits', 'reporters'])
  })

  it('enforces navigation access from required permissions', () => {
    const settingsOnly = principalWith(['settings.read'], ['Staff'])
    expect(canAccessNavigationPath(settingsOnly, '/settings')).toBe(true)
    const settingsWriter = principalWith(['settings.write'], ['Staff'])
    expect(canAccessNavigationPath(settingsWriter, '/settings')).toBe(true)
    expect(canAccessNavigationPath(settingsOnly, '/users')).toBe(false)

    const usersReader = principalWith(['users.read'])
    expect(canAccessNavigationPath(usersReader, '/users')).toBe(true)
    expect(canAccessNavigationPath(usersReader, '/audit')).toBe(false)

    const serversReader = principalWith(['servers.read'], ['Staff'])
    expect(canAccessNavigationPath(serversReader, '/servers')).toBe(true)
    expect(canAccessNavigationPath(serversReader, '/observability')).toBe(false)
    expect(canAccessNavigationPath(serversReader, '/scheduler')).toBe(false)
    expect(canAccessNavigationPath(serversReader, '/documentation')).toBe(true)
  })

  it('blocks module paths when effective config disables a module', () => {
    resetEffectiveModuleEnablement()
    applyEffectiveModuleEnablement({
      modules: [
        {
          moduleId: 'dashboard',
          flagKey: 'modules.dashboard.enabled',
          enabled: true,
          enabledByDefault: true,
          source: 'default',
        },
        {
          moduleId: 'administration',
          flagKey: 'modules.administration.enabled',
          enabled: false,
          enabledByDefault: true,
          source: 'config',
        },
        {
          moduleId: 'settings',
          flagKey: 'modules.settings.enabled',
          enabled: true,
          enabledByDefault: true,
          source: 'default',
        },
      ],
    })

    expect(isPathModuleEnabled('/users')).toBe(false)
    expect(canAccessNavigationPath(principalWith(['users.read']), '/users')).toBe(false)
    expect(isPathModuleEnabled('/servers')).toBe(true)
    resetEffectiveModuleEnablement()
  })
})

File: backend/internal/moduleenablement/registry_test.go
package moduleenablement

import "testing"

func TestResolveEffectiveUsesDefaultsAndConfigOverrides(t *testing.T) {
	items := ResolveEffective(
		map[string]bool{"modules.settings.enabled": false},
		map[string]bool{"modules.administration.enabled": false},
	)

	if len(items) != 12 {
		t.Fatalf("expected 12 module definitions, got %d", len(items))
	}

	if items[0].ModuleID != "dashboard" || !items[0].Enabled || items[0].Source != "default" || items[0].Editable {
		t.Fatalf("unexpected dashboard effective config: %+v", items[0])
	}

	if items[1].ModuleID != "administration" || items[1].Enabled || items[1].Source != "runtime" || !items[1].Editable {
		t.Fatalf("expected administration module to be disabled by runtime override: %+v", items[1])
	}

	if items[2].ModuleID != "settings" || items[2].Enabled || items[2].Source != "config" || items[2].Editable {
		t.Fatalf("expected settings module to be disabled by config override: %+v", items[2])
	}

	for _, moduleID := range []string{"servers", "requests", "deliveries", "jobs", "scheduler", "observability", "documentation", "orgunits", "reporters"} {
		found := false
		for _, item := range items {
			if item.ModuleID != moduleID {
				continue
			}
			found = true
			if !item.Enabled || item.Source != "default" || !item.Editable {
				t.Fatalf("expected %s module enabled by default and runtime-editable: %+v", moduleID, item)
			}
		}
		if !found {
			t.Fatalf("expected module definition for %s", moduleID)
		}
	}
}

func TestValidateOverridesRejectsUnknownKeys(t *testing.T) {
	err := ValidateOverrides(map[string]bool{
		"modules.unknown.enabled": false,
	})
	if err == nil {
		t.Fatal("expected unknown key validation error")
	}

	if err := ValidateOverrides(map[string]bool{
		"modules.dashboard.enabled": true,
	}); err != nil {
		t.Fatalf("expected known key to pass validation: %v", err)
	}
}
